File: csharp/src/AutomationRoot/Automation.Extensions/Components/WebDriverFactory.cs
using Automation.Extensions.Contracts;
using Newtonsoft.Json;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Edge;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.IE;
using OpenQA.Selenium.Remote;
using System;
using System.Collections.Generic;
using System.Text;

namespace Automation.Extensions.Components
{
    public class WebDriverFactory
    {
        private readonly DriverParams driverParams;

        public WebDriverFactory(string driverParamsJson)
            : this(LoadParams(driverParamsJson)) { }

        public WebDriverFactory(DriverParams driverParams)
        {
            this.driverParams = driverParams;
            if (string.IsNullOrEmpty(driverParams.Binaries) || driverParams.Binaries == ".")
            {
                driverParams.Binaries = Environment.CurrentDirectory;
            }
        }

        /// <summary>
        /// generates web-driver based on input params
        /// </summary>
        /// <returns>web-driver instance</returns>
        public IWebDriver Get()
        {
            if (!string.Equals(driverParams.Source, "REMOTE", StringComparison.OrdinalIgnoreCase))
            {
                return GetDriver();
            }
            return GetRemoteDriver();
        }

        // local web-drivers
        private IWebDriver GetChrome() => new ChromeDriver(driverParams.Binaries);
        private IWebDriver GetFirefox() => new FirefoxDriver(driverParams.Binaries);
        private IWebDriver GetInternetExplorer() => new InternetExplorerDriver(driverParams.Binaries);
        private IWebDriver GetEdge() => new EdgeDriver(driverParams.Binaries);

        private IWebDriver GetDriver()
        {
            switch (driverParams.Driver.ToUpper())
            {
                case "EDGE": return GetEdge();
                case "IE": return GetInternetExplorer();
                case "FIREFOX": return GetFirefox();
                case "CHROME":
                default: return GetChrome();
            }
        }

        // remote web-drivers
        private IWebDriver GetRemoteChrome()
            => new RemoteWebDriver(new Uri(driverParams.Binaries), new ChromeOptions());

        private IWebDriver GetRemoteFirefox()
            => new RemoteWebDriver(new Uri(driverParams.Binaries), new FirefoxOptions());

        private IWebDriver GetRemoteInternetExplorer()
            => new RemoteWebDriver(new Uri(driverParams.Binaries), new InternetExplorerOptions());

        private IWebDriver GetRemoteEdge()
            => new RemoteWebDriver(new Uri(driverParams.Binaries), new EdgeOptions());

        private IWebDriver GetRemoteDriver()
        {
            switch (driverParams.Driver.ToUpper())
            {
                case "EDGE": return GetRemoteEdge();
                case "IE": return GetRemoteInternetExplorer();
                case "FIREFOX": return GetRemoteFirefox();
                case "CHROME":
                default: return GetRemoteChrome();
            }
        }

        // load json into driver-params object
        private static DriverParams LoadParams(string driverParamsJson)
        {
            if (string.IsNullOrEmpty(driverParamsJson))
            {
                return new DriverParams { Source = "Local", Driver = "Chrome", Binaries = "." };
            }
            return JsonConvert.DeserializeObject<DriverParams>(driverParamsJson);
        }
    }
}

File: java/src/automation-root/automation-extensions/src/main/java/com/automation/extensions/components/WebDriverFactory.java
package com.automation.extensions.components;

import com.automation.extensions.contracts.DriverParams;
import com.google.gson.Gson;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeDriverService;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.edge.EdgeDriver;
import org.openqa.selenium.edge.EdgeDriverService;
import org.openqa.selenium.edge.EdgeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.firefox.GeckoDriverService;
import org.openqa.selenium.ie.InternetExplorerDriver;
import org.openqa.selenium.ie.InternetExplorerDriverService;
import org.openqa.selenium.ie.InternetExplorerOptions;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Path;
import java.nio.file.Paths;

public class WebDriverFactory {
    private final DriverParams driverParams;

    public WebDriverFactory(String driverParamsJson){
        this(loadParams(driverParamsJson));
    }

    public WebDriverFactory (DriverParams driverParams){
        this.driverParams = driverParams;
    }

    /**
     * generate web-driver based on input params
     * @return web-driver instance
     * @throws MalformedURLException from inner factory
     */
    public WebDriver get() throws MalformedURLException {
        // setup
        boolean isBinaries = driverParams.getBinaries() != null && driverParams.getBinaries().length() > 0;
        boolean isRemote = isBinaries && driverParams.getBinaries().toUpperCase().startsWith("HTTP");

        // get
        if (!isRemote) {
            return getDriver();
        }
        return getRemoteDriver();
    }

    // local web-drivers
    private WebDriver getChrome() {
        Path base = Paths.get(driverParams.getBinaries());
        Path path = base.resolve("chromedriver.exe");

        // setup options
        ChromeOptions chromeOptions = new ChromeOptions();
        chromeOptions.addArguments("--remote-allow-origins=*");

        // CHROME
        ChromeDriverService chromeDriverService = new ChromeDriverService
                .Builder()
                .usingDriverExecutable(new File(path.toString()))
                .usingAnyFreePort()
                .build();

        return new ChromeDriver(chromeDriverService, chromeOptions);
    }

    private WebDriver getFirefox() {
        Path base = Paths.get(driverParams.getBinaries());
        Path path = base.resolve("geckodriver.exe");

        // FIREFOX
        GeckoDriverService geckoDriverService = new GeckoDriverService
                .Builder()
                .usingDriverExecutable(new File(path.toString()))
                .usingAnyFreePort()
                .build();

        return new FirefoxDriver(geckoDriverService);
    }

    private WebDriver getInternetExplorer() {
        Path base = Paths.get(driverParams.getBinaries());
        Path path = base.resolve("IEDriverServer.exe");

        // INTERNET EXPLORER
        InternetExplorerDriverService ieDriverService = new InternetExplorerDriverService
                .Builder()
                .usingDriverExecutable(new File(path.toString()))
                .usingAnyFreePort()
                .build();

        return new InternetExplorerDriver(ieDriverService);
    }

    private WebDriver getEdge() {
        // setup
        Path base = Paths.get(driverParams.getBinaries());
        Path path = base.resolve("msedgedriver.exe");

        // setup options
        EdgeOptions edgeOptions = new EdgeOptions();
        edgeOptions.addArguments("--remote-allow-origins=*");

        // EDGE
        EdgeDriverService edgeDriverService = new EdgeDriverService
                .Builder()
                .usingDriverExecutable(new File(path.toString()))
                .usingAnyFreePort()
                .build();

        // get
        return new EdgeDriver(edgeDriverService, edgeOptions);
    }

    private WebDriver getDriver(){
        switch (driverParams.getDriver().toUpperCase()){
            case "EDGE": return getEdge();
            case "IE": return getInternetExplorer();
            case "FIREFOX": return getFirefox();
            case "CHROME":
            default: return getChrome();
        }
    }

    // remote web-drivers
    private WebDriver getRemoteChrome() throws MalformedURLException {
        return new RemoteWebDriver(new URL(driverParams.getBinaries()), new ChromeOptions());
    }

    private WebDriver getRemoteFirefox() throws MalformedURLException {
        return new RemoteWebDriver(new URL(driverParams.getBinaries()), new FirefoxOptions());
    }

    private WebDriver getRemoteInternetExplorer() throws MalformedURLException {
        return new RemoteWebDriver(new URL(driverParams.getBinaries()), new InternetExplorerOptions());
    }

    private WebDriver getRemoteEdge() throws MalformedURLException {
        return new RemoteWebDriver(new URL(driverParams.getBinaries()), new EdgeOptions());
    }

    private WebDriver getRemoteDriver() throws MalformedURLException {
        switch (driverParams.getDriver().toUpperCase()){
            case "EDGE": return getRemoteEdge();
            case "IE": return getRemoteInternetExplorer();
            case "FIREFOX": return getRemoteFirefox();
            case "CHROME":
            default: return getRemoteChrome();
        }
    }

    // load json into driver-params object
    private static DriverParams loadParams(String driverParamsJson){
        if(driverParamsJson == null || driverParamsJson.isEmpty()){
            return new DriverParams().setDriver("Chrome").setBinaries(".");
        }
        return new Gson().fromJson(driverParamsJson, DriverParams.class);
    }
}
